ENH: More accurate signal.freqresp for zpk systems

e-q opened the pull request:
This PR addresses part of #5648, making higher order zpk frequency responses
more accurate. The SOS part still needs analog zpk<->sos conversion (see
gh-5668).

This consists of:
- Modifying `findfreqs` to accept zeros and poles directly
- A new function `freqs_zpk`, analogous to `freqs`
- A new function `_polyrootval`, which evaluates the s-plane polynomials
  without the lossy polynomial expansion. (I plan to submit a PR with this
  function to numpy, but it would be a while before we can expect users to have
  it, I imagine)
- New logic in `freqresp` that detects when the input is in ZPK form, and calls
  `freqs_zpk` if appropriate. (SS systems still go to tf, since we don't have a
  direct SS<->zpk translator [gh-5912])
- Tests:
  - Basic tests for `freqs`, which were nonexistent
  - Basic tests for `freqs_zpk`
  - New test for `freqresp` on ZPK input
  - Consolidation of some redundant `freqresp` tests

e-q commented:
Here's an example of the difference between `freqs` and `freqs_zpk`

```python
N=150
sysTF=sig.lti(*sig.cheby1(N, 6, np.pi, btype='lowpass', analog=True, output='ba'))
sysZPK=sig.lti(*sig.cheby1(N,6, np.pi, btype='lowpass', analog=True, output='zpk'))
w, h = sysTF.freqresp()
w2,h2 = sysZPK.freqresp()
plt.loglog(w,np.abs(h),w2,np.abs(h2))
plt.ylim([1e-2,10])
plt.legend(['ba','zpk''])
```
![freqs_zpk_vs_freqs](https://cloud.githubusercontent.com/assets/6894112/14477919/1d8a9884-00c6-11e6-9eb8-7cb3d22195c5.png)

e-q commented:
Third time's a charm to get the doctests to pass...

e-q commented:
Updated given feedback from numpy/numpy#7542

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -34,18 +34,22 @@ class BadCoefficients(UserWarning):
 
 
-def findfreqs(num, den, N):
+def findfreqs(num, den, N, input='ba'):
```
adding an `input=` parameter is another way to deal with https://github.com/scipy/scipy/issues/3259
but I would prefer making the function accept a system as the argument instead.
`findfreqs` isn't a great name to begin with, so can we think of a better name and call it like `important_filter_frequencies(system, N)` instead?

e-q replied:
I agree, but its maybe outside of the scope of this PR. I think it needs to happen in a unified way when we make this kind of change across the board.

rgommers replied:
I agree that gh-3259 is out of scope here, and adding a keyword to select ``'zpk'`` isn't going to get in the way of a new API, whatever we choose in gh-3259.

rgommers replied:
``input`` is a bit unfortunate as a keyword though, given that it's a builtin.

Eric89GXL replied:
Maybe "kind" is better

endolith commented on scipy/signal/ltisys.py:
```diff
@@ -1744,12 +1744,15 @@ def freqresp(system, w=None, n=10000):
+    if isinstance(system, (TransferFunction, ZerosPolesGain)):
         sys = system
     elif isinstance(system, lti):
         sys = system.to_tf()
```
shouldn't this be to_zpk now?

e-q replied:
In principle, yes. However, right now, to get to this line the system must be state space, which means `to_zpk` is actually `system.to_tf().to_zpk()`, which doesn't win anything, so I didn't include it.

Though, if we put `to_zpk` here, then it's less likely we'll forget to change this when a proper `ss2zpk` is written, so I do agree it should change.

rgommers commented on scipy/signal/filter_design.py:
```diff
@@ -161,6 +171,147 @@ def freqs(b, a, worN=None, plot=None):
+
+    Notes
+    -----
+    Will be included in numpy's polynomial submodule in v1.12
```
Better to add this function in ``scipy/_lib/_numpy_compat.py`` then I think.

e-q commented:
Thanks for the feedback! I've changed the keyword to `kind`, and moved the polynomial root evaluation to `_numpy_compat`. (Which I'm glad to learn about)

If the tests pass, this is wrapped up on my end.

e-q commented:
Just kidding. Had to rebase given #4881. This prompted me to realize that I can extend this to `freqz` as well. New commits forthcoming...

Eric89GXL commented:
If you're extending `freqz` to `zpk` can you also do it for SOS (#4465)?

endolith commented on scipy/_lib/_numpy_compat.py:
```diff
@@ -72,3 +72,94 @@ def assert_raises_regex(exception_class, expected_regexp,
+        >>> a = np.arange(4).reshape(2,2)
+        >>> a
+        array([[0, 1],
+            [2, 3]])
```
should these be indented more?

e-q replied:
Good catch! Poor copy-pasting on my part...

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -161,6 +173,94 @@ def freqs(b, a, worN=None, plot=None):
     return w, h
 
 
+def freqs_zpk(z, p, k, worN=None, plot=None):
```
Does this really need the `plot=` parameter?  It isn't really useful

endolith replied:
Also does this function need to be public if `freqresp` provides a better interface to do the same thing?

tacaswell replied:
Might be better as a paramterized factory function?

endolith replied:
I mean that `freqs` and `freqz` should only exist for backwards/matlab compatibility and the New Way should be `freqresp` and `dfreqresp`.

Eric89GXL replied:
I'm fine with leaving it in for consistency or removing it (as I wish it weren't in `freqs` in the first place)

e-q commented:
Good idea @Eric89GXL, I can probably have that by this weekend. 

@endolith, I agree that `(d)freqresp` should be encouraged, and am ok with these functions being private. However the SOS versions will have to be public, since we don't have system objects for them yet.

endolith commented:
> However the SOS versions will have to be public, since we don't have system objects for them yet.

Well you can still call `w, h = dfreqresp(sos_array)` without a dedicated LTI object, no?

e-q commented:
Currently, if `dfreqresp` is passed something other than a `dlti` object, it calls `dlti` on the arguments. So, to make it work with an `sos_array` input, we'd either have to create a SOS `dlti` subclass, or hack on some SOS-detecting logic.

The former is out of scope of this PR, while the latter doesn't appeal to me too much.

endolith commented:
> we'd either have to create a SOS dlti subclass

(For the record, this might be more useful as a generic multi-stage filter class, with arbitrary orders for each section, rather than only biquads.)

rgommers commented:
needs a rebase.

this is pretty much done right, if SOS is left to a separate PR?

e-q commented:
Thanks for the reminder! Rebased, and I think it is indeed ready for review.

rgommers commented:
TravisCI failure is a single PEP8 issue:

    $ pep8 scipy benchmarks/benchmarks
    scipy/signal/tests/test_filter_design.py:17:69: E203 whitespace before ','
    E203 whitespace before ','

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -260,6 +360,101 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
+    h : ndarray
+        The frequency response.
+
+    Notes
```
See Also section would be nice

e-q commented:
Time marches on: rebased again

No one spoke in favor of the `plot` kwargs, so I've removed them from the new functions. This should be ready to go.

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -162,6 +173,84 @@ def freqs(b, a, worN=None, plot=None):
+    See Also
+    --------
+    freqs : Compute the frequency response of an analog filter in TF form
+    freqz : Compute the frequency response of a digital filter.
```
might as well add `freqz_zpk`

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -162,6 +173,84 @@ def freqs(b, a, worN=None, plot=None):
+    frequency response::
+
+                (jw-z[0]) * (jw-z[1]) * ... * (jw-z[-1])
+     H(w) = k * -------------------------------------------------------
```
this doesn't need to be as long

Eric89GXL commented on scipy/signal/filter_design.py:
```diff
@@ -260,6 +349,98 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
+    See Also
+    --------
+    freqs : Compute the frequency response of an analog filter in TF form
+    freqz : Compute the frequency response of a digital filter.
```
freqs_zpk

e-q commented:
Thanks for the feedback @Eric89GXL, we should be all set.

Eric89GXL commented:
@endolith do you have time to take another look to see if you're happy with this?

e-q commented:
Two weeks with no objections; I'm going to go ahead and merge this so it doesn't get too out of date.

rgommers commented:
Nice, thanks Eric

endolith commented:
Wait, how did we end up with both `freqz_zpk` and `sosfreqz`?  :/  Can we add an alias of either `zpkfreqz` or `freqz_sos` to harmonize them?

e-q commented:
Yeah, the naming is unfortunate. I was kind of hoping that with 1.0 we could make a clean break with the naming of some of the older stuff and harmonize the whole LTI design/analysis/filtering situation, but I've been buried under my thesis for a while.
